src/lve_texture.cpp

```diff
@@ -0,0 +1,337 @@
+      mTextureImage,
+      mTextureImageMemory);
+
+  VkImageViewCreateInfo viewInfo{};
+  viewInfo.sType = VK_STRUCTURE_TYPE_IMAGE_VIEW_CREATE_INFO;
+  viewInfo.viewType = VK_IMAGE_VIEW_TYPE_2D;
+  viewInfo.format = format;
+  viewInfo.subresourceRange = {};
+  viewInfo.subresourceRange.aspectMask = aspectMask;
+  viewInfo.subresourceRange.baseMipLevel = 0;
+  viewInfo.subresourceRange.levelCount = 1;
+  viewInfo.subresourceRange.baseArrayLayer = 0;
+  viewInfo.subresourceRange.layerCount = 1;
+  viewInfo.image = mTextureImage;
+  if (vkCreateImageView(device.device(), &viewInfo, nullptr, &mTextureImageView) != VK_SUCCESS) {
+    throw std::runtime_error("failed to create texture image view!");
+  }
+
+  // Sampler should be seperated out
+  if (usage & VK_IMAGE_USAGE_SAMPLED_BIT) {
+    // Create sampler to sample from the attachment in the fragment shader
+    VkSamplerCreateInfo samplerInfo{};
+    samplerInfo.sType = VK_STRUCTURE_TYPE_SAMPLER_CREATE_INFO;
+    samplerInfo.magFilter = VK_FILTER_LINEAR;
+    samplerInfo.minFilter = VK_FILTER_LINEAR;
+    samplerInfo.mipmapMode = VK_SAMPLER_MIPMAP_MODE_LINEAR;
+    samplerInfo.addressModeU = VK_SAMPLER_ADDRESS_MODE_CLAMP_TO_BORDER;
+    samplerInfo.addressModeV = samplerInfo.addressModeU;
+    samplerInfo.addressModeW = samplerInfo.addressModeU;
+    samplerInfo.mipLodBias = 0.0f;
+    samplerInfo.maxAnisotropy = 1.0f;
+    samplerInfo.minLod = 0.0f;
+    samplerInfo.maxLod = 1.0f;
+    samplerInfo.borderColor = VK_BORDER_COLOR_FLOAT_OPAQUE_BLACK;
+
+    if (vkCreateSampler(device.device(), &samplerInfo, nullptr, &mTextureSampler) != VK_SUCCESS) {
+      throw std::runtime_error("failed to create sampler!");
+    }
+
+    VkImageLayout samplerImageLayout = imageLayout == VK_IMAGE_LAYOUT_COLOR_ATTACHMENT_OPTIMAL
+                                           ? VK_IMAGE_LAYOUT_SHADER_READ_ONLY_OPTIMAL
+                                           : VK_IMAGE_LAYOUT_DEPTH_STENCIL_READ_ONLY_OPTIMAL;
+    mDescriptor.sampler = mTextureSampler;
+    mDescriptor.imageView = mTextureImageView;
+    mDescriptor.imageLayout = samplerImageLayout;
+  }
+}
+
+LveTexture::~LveTexture() {
+  vkDestroySampler(mDevice.device(), mTextureSampler, nullptr);
+  vkDestroyImageView(mDevice.device(), mTextureImageView, nullptr);
+  vkDestroyImage(mDevice.device(), mTextureImage, nullptr);
+  vkFreeMemory(mDevice.device(), mTextureImageMemory, nullptr);
+}
+
+std::unique_ptr<LveTexture> LveTexture::createTextureFromFile(
+    LveDevice &device, const std::string &filepath) {
+  return std::make_unique<LveTexture>(device, filepath);
+}
+
+void LveTexture::updateDescriptor() {
+  mDescriptor.sampler = mTextureSampler;
+  mDescriptor.imageView = mTextureImageView;
+  mDescriptor.imageLayout = mTextureLayout;
+}
+
+void LveTexture::createTextureImage(const std::string &filepath) {
+  int texWidth, texHeight, texChannels;
+  // stbi_set_flip_vertically_on_load(1);  // todo determine why texture coordinates are flipped
```

Comment:
https://stackoverflow.com/questions/19770296/should-i-vertically-flip-the-lines-of-an-image-loaded-with-stb-image-to-use-in-o


use 
```cpp
stbi_set_flip_vertically_on_load(1);
```